R/preprocessing.R

```diff
@@ -1838,22 +1838,62 @@ NormalizeData.Seurat <- function(
 #' @param k  The rank of the rank-k approximation. Set to NULL for automated choice of k.
 #' @param q  The number of additional power iterations in randomized SVD when
 #' computing rank k approximation. By default, q=10.
+#' @param quantile.prob The quantile probability to use when calculating threshold.
+#' By default, quantile.prob = 0.001.
+#' @param use.mkl Use the Intel MKL based implementation of SVD. Needs to be
+#' installed from https://github.com/KlugerLab/rpca-mkl.
+#' @param mkl.seed Only relevant if use.mkl=T. Set the seed for the random
+#' generator for the Intel MKL implementation of SVD. Any number <0 will
+#' use the current timestamp. If use.mkl=F, set the seed using
+#' set.seed() function as usual.
 #'
 #' @rdname RunALRA
 #' @export
 #'
-RunALRA.default <- function(object, k = NULL, q = 10, ...) {
+RunALRA.default <- function(
+  object,
+  k = NULL,
+  q = 10,
+  quantile.prob = 0.001,
+  use.mkl = FALSE,
+  mkl.seed = -1,
+  ...
+) {
   CheckDots(...)
   A.norm <- t(x = as.matrix(x = object))
   message("Identifying non-zero values")
   originally.nonzero <- A.norm > 0
   message("Computing Randomized SVD")
-  fastDecomp.noc <- rsvd(A = A.norm, k = k, q = q)
+  if (use.mkl) {
+    fastDecomp.noc <- setNames(
+      object = vector(mode = "list", length = 3),
+      nm = c("u", "d", "v")
+    )
+    fastPCAOut <- fastRPCA::fastPCA(
```

Comment:
[fastRPCA](https://github.com/KlugerLab/rpca-mkl) is not on CRAN, and it is still under active development. This is mainly because of its dependence on Intel MKL. In order to not require users to install Intel MKL on their computers, we have to distribute a binary...and that makes it difficult to submit to CRAN. 

ALRA does not depend on fastRPCA; it is only used if the user sets the `use.mkl` argument indicating they installed this package from the Github repository. Given that it will not interfere with normal usage (i.e. no one will be required to install this package), is it still not allowed? From your comment I understand that to be the case, but I want to confirm before we remove it.

Reply:
Unfortunately, that is correct. If we use a package in Seurat, CRAN requires that it be hosted on CRAN, Bioconductor, or another CRAN-like repository. Using a package hosted only on GitHub is not allowed for CRAN-hosted packages, and if fastRPCA isn't going to CRAN anytime soon, we do ask that it be removed from this PR.

However, we can include it in SeuratWrappers; that package will not be going to CRAN as it was designed to allow `Seurat` object-based methods using packages not hosted on CRAN, Bioconductor, or another CRAN-like repository.

Reply:
> Unfortunately, that is correct. If we use a package in Seurat, CRAN requires that it be hosted on CRAN, Bioconductor, or another CRAN-like repository. Using a package hosted only on GitHub is not allowed for CRAN-hosted packages, and if fastRPCA isn't going to CRAN anytime soon, we do ask that it be removed from this PR.

That makes sense. 

> However, we can include it in SeuratWrappers; that package will not be going to CRAN as it was designed to allow Seurat object-based methods using packages not hosted on CRAN, Bioconductor, or another CRAN-like repository.

Given this, before I go ahead and remove the fastRPCA compatibility, I was wondering about your timeframe for moving ALRA to SeuratWrappers. Since we can include this functionality after it is moved to SeuratWrappers---and you are planning on moving it anyway---perhaps we should just wait until then to update ALRA in Seurat?

Reply:
We already have an implementation (basically copy-pasted from Seurat) [available in SeuratWrappers](https://github.com/satijalab/seurat-wrappers/blob/master/R/alra.R). We are currently preparing Seurat v3.1.0, in which `RunALRA` will be marked as deprecated. We will likely mark it as defunct in a later point release of the v3.1.X series (somewhere around v3.1.1, v3.1.2, or v3.1.3) or at the latest v3.2.0, at which point `RunALRA` in Seurat will cease to work and users will be directed to SeuratWrappers.

Since we already have an implementation there, we can take your updates here, including usage of fastRPCA, and incorporate them into the SeuratWrappers implementation of `RunALRA`. In addition, we can add everything except fastRPCA usage here, with a message if `use.mkl = TRUE` directing users to `RunALRA` in SeuratWrappers.